### ISSUE TITLE
Using 'extracting' with InstanceOfAssertFactory will lead to an unexpected AssertionError if extraced value is null 

### ISSUE THREAD

**Issue by kamil-sita** [assertj/assertj#3266]
**Describe the bug**
AssertJ allows extracting values from the original object using "extracting" method in order to test them in a more fluent way. AssertJ also allows using InstanceOfAssertFactory in "extracting" to specify which assertions should be used to test the extracted value.

However, there's an unexpected non-null check when using InstanceOfAssertFactory, which leads to an exception being thrown if the extracted value is null. This is not the case when not using InstanceOfAssertFactory, see the example below:

```java
import org.assertj.core.api.InstanceOfAssertFactories;
import org.testng.annotations.Test;

import static org.assertj.core.api.Assertions.assertThat;

public class Reproduced {

    @Test
    public void should_testNullCorrectly() {
        // given
        StringHolder holder = new StringHolder(null);

        // then
        assertThat(holder)
            .extracting(StringHolder::value)
            .isNull(); // this test will succeed
    }

    @Test
    public void should_testNullCorrectlyWithAssertFactory() {
        // given
        StringHolder holder = new StringHolder(null);

        // then
        assertThat(holder)
            .extracting(StringHolder::value, InstanceOfAssertFactories.STRING)
            .isNull(); // this test will fail
    }

    public record StringHolder(String value) { }
}
```

* assertj core version: tested with 3.16.1 and 3.24.2
* java version: 17.0.6
* test framework version: tested with TestNG 6.9.10 and JUnit 5.9.2

**Test case reproducing the bug**

```java

import org.assertj.core.api.InstanceOfAssertFactories;
import org.testng.annotations.Test;

import static org.assertj.core.api.Assertions.assertThat;

public class Reproduced {

    @Test
    public void should_testNullCorrectlyWithAssertFactory() {
        // given
        StringHolder holder = new StringHolder(null);

        // then
        assertThat(holder)
            .extracting(StringHolder::value, InstanceOfAssertFactories.STRING)
            .isNull(); // this test will fail
    }

    public record StringHolder(String value) { }
}
```

**Comment by scordio** [comment]
[`extracting(Function, InstanceOfAssertFactory)`](https://www.javadoc.io/doc/org.assertj/assertj-core/latest/org/assertj/core/api/AbstractObjectAssert.html#extracting(java.util.function.Function,org.assertj.core.api.InstanceOfAssertFactory)) internally relies on [`asInstanceOf`](https://www.javadoc.io/doc/org.assertj/assertj-core/latest/org/assertj/core/api/Assert.html#asInstanceOf(org.assertj.core.api.InstanceOfAssertFactory)) which performs a preliminary [`isInstanceOf`](https://www.javadoc.io/doc/org.assertj/assertj-core/latest/org/assertj/core/api/Assert.html#isInstanceOf(java.lang.Class)) check, the source of the `AssertionError` you are experiencing.

Is there a reason why you are using `InstanceOfAssertFactories.STRING` even though you already expect the value to be null?

On a side note, the Javadoc of `extracting` and `asInstanceOf` could be improved to cover such cases.

**Comment by kamil-sita** [comment]
> Is there a reason why you are using `InstanceOfAssertFactories.STRING` even though you already expect the value to be null?

Not exactly, no. I have created helper methods for exporting properties of the objects I'm testing, like this:

```java
public AbstractStringAssert<?> assertionForPropertyXyz() {
   // some complicated code I would prefer not to repeat tens of times
   
   return assertThat(abc)
       .extracting(Abc::getXyz, InstanceOfAssertFactories.STRING);
}
```

and have planned to use them in my code like so:

```java
// test 1
   someObject.assertionForPropertyXyz().startsWith("lorem ipsum");
// test 2
   someObject.assertionForPropertyXyz().isNull();
```

I realize that it might be an uncommon approach, however I also think that having ``extracting`` methods behave differently here is very surprising. Personally I think that both should behave in more-or-less the same way.

**Comment by kamil-sita** [comment]
Thank you!

**Comment by kamil-sita** [comment]
Hi, I wonder if you had an occasion to discuss this issue internally?

**Comment by bekoenig** [comment]
Hi again,

I refactored from  using `asInstanceOf` to [`extracting` using `assertFactory`](https://www.javadoc.io/doc/org.assertj/assertj-core/latest/org/assertj/core/api/AbstractAssert.html#extracting(java.lang.String,org.assertj.core.api.AssertFactory)) to fix this issue. But this workaround is only for custom assertions possible which extends from `AbstractAssert`.

Greetings,
Ben

**Comment by scordio** [comment]
Hi everyone, the rough direction we have in mind is to introduce an `Optional::map`-like alternative to `extracting`. This way, further transformations like `asInstanceOf` can be chained, but only assertion methods would "materialize" the result without causing intermediate NPEs.

We haven't had the chance to investigate this direction deeply, but I'll schedule this topic for further investigation in one of the upcoming v4 milestones.

**Comment by scordio** [comment]
As discussed internally, we will deliver this in 3.28 and also backport https://github.com/assertj/assertj/commit/7c657bc44afcc4ad67a4aba8fb16f801de2a3656 as part of the fix.
